napi-inl.h

```diff
@@ -865,47 +942,65 @@ inline const T* TypedArray_<T,A>::Data() const {
 // Function class
 ////////////////////////////////////////////////////////////////////////////////
 
-inline Function Function::New(napi_env env,
-                              VoidFunctionCallback cb,
-                              const char* utf8name,
-                              void* data) {
-  // TODO: Delete when the function is destroyed
-  CallbackData* callbackData = new CallbackData(cb, data);
+// Helpers to handle functions exposed from C++.
+namespace helper {
 
-  // TODO: set the function name
-  napi_value value;
-  napi_status status = napi_create_function(
-    env, utf8name, VoidFunctionCallbackWrapper, callbackData, &value);
-  if (status != napi_ok) throw Error::New(env);
-  return Function(env, value);
-}
+template <typename Callable, typename Return>
+struct CallbackData {
+  static inline
+  napi_value Wrapper(napi_env env, napi_callback_info info) {
+    try {
+      CallbackInfo callbackInfo(env, info);
+      CallbackData* callbackData =
+        static_cast<CallbackData*>(callbackInfo.Data());
+      return callbackData->callback(callbackInfo);
+    }
+    NAPI_RETHROW_JS_ERROR(env)
+  }
+
+  Callable callback;
+};
+
+template <typename Callable>
+struct CallbackData<Callable, void> {
+  static inline
+  napi_value Wrapper(napi_env env, napi_callback_info info) {
+    try {
+      CallbackInfo callbackInfo(env, info);
+      CallbackData* callbackData =
+        static_cast<CallbackData*>(callbackInfo.Data());
+      callbackData->callback(callbackInfo);
+    }
+    NAPI_RETHROW_JS_ERROR(env)
+  }
 
+  Callable callback;
+};
+
+}  // namespace helper
+
+template<typename Callable>
 inline Function Function::New(napi_env env,
-                              FunctionCallback cb,
-                              const char* utf8name,
-                              void* data) {
```

Comment:
>  `auto data = info.Data().real_data;`
>  `info.Data().call(real_data);`

Sorry, what is `real_data` in that context?

And also: I have no problem re-adding a generic data value here, whether that’s a `Value` or a pointer if you would prefer that. I just decided to leave it away for now because it didn’t work until now, and allowing `std::function`s more or less circumvents the problem solved by that extra opaque anyway.